pkg/store/downtime_store.go

```diff
@@ -0,0 +1,350 @@
+	if err != nil {
+		return nil, fmt.Errorf("error while validating external ID: %w", err)
+	}
+
+	if len(existing_id) > 0 && existing_id != st.ID {
+		return nil, errors.New("could not patch downtime window: external ID conflicts with existing record")
+	}
+
+	return s.updateWindow(&st)
+}
+
+func (s *DowntimeStore) validate(w *dbDowntimeWindow) error {
+	if w.StartTime <= 0 {
+		return errors.New("validation error: start time must be set")
+	}
+	if w.EndTime > 0 && w.StartTime > w.EndTime {
+		return errors.New("validation error: end time must be after start time")
+	}
+	return nil
+}
+
+func (s *DowntimeStore) updateWindow(w *dbDowntimeWindow) (*types.DowntimeWindow, error) {
+	q := `UPDATE downtime SET id = :id, start_time = :start_time,  end_time = :end_time, title = :title, description = :description, external_id = :external_id, external_link = :external_link, affects = :affects WHERE id == :id`
+	_, err := s.db.NamedExec(q, w)
+	if err != nil {
+		return nil, fmt.Errorf("unable to update downtime window: %w", err)
+	}
+
+	rv, err := convertFromDbStruct(w)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert update result: %w", err)
+	}
+
+	return rv, nil
+}
+
+func (s *DowntimeStore) getWindowById(id string) (dbDowntimeWindow, error) {
+	result := dbDowntimeWindow{}
+	err := s.db.Get(&result, "SELECT * FROM downtime WHERE id == ?", id)
+	if err != nil {
+		return dbDowntimeWindow{}, fmt.Errorf("error while querying record by ID: %w", err)
+	}
+	return result, nil
+}
+
+func (s *DowntimeStore) idFromExternalID(externalID string) (string, error) {
+	if len(externalID) == 0 {
+		//NOTE(aa): empty externalIDs are not considered
+		return "", nil
+	}
+
+	results := []dbDowntimeWindow{}
+	err := s.db.Select(&results, "SELECT * FROM downtime WHERE external_id == ? LIMIT 1", externalID)
+	if err != nil {
+		return "", fmt.Errorf("error while querying for external ID: %w", err)
+	}
+
+	if len(results) > 0 {
+		return results[0].ID, nil
+	}
+	return "", nil
+}
+
+func convertToDbStruct(w *types.DowntimeWindow) (dbDowntimeWindow, error) {
+	affects, err := json.Marshal(w.Affects)
+	if err != nil {
+		return dbDowntimeWindow{}, fmt.Errorf("could not convert downtime window: %w", err)
+	}
+	nw := dbDowntimeWindow{
+		ID:           w.ID,
+		Title:        w.Title,
+		Description:  w.Description,
+		ExternalID:   w.ExternalID,
+		ExternalLink: w.ExternalLink,
+		Affects:      string(affects),
+	}
+
+	if w.StartTime != nil {
+		nw.StartTime = w.StartTime.Unix()
+	}
+	if w.EndTime != nil {
+		nw.EndTime = w.EndTime.Unix()
+	}
+
+	if len(nw.ID) == 0 {
+		nw.ID = uuid.New().String()
+	}
+
+	return nw, nil
+}
+
+func convertFromDbStruct(w *dbDowntimeWindow) (*types.DowntimeWindow, error) {
+	affects := []types.AffectedClusterMatcher{}
+	err := json.Unmarshal([]byte(w.Affects), &affects)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert downtime window: %w", err)
+	}
+	st, en := time.Unix(w.StartTime, 0), time.Unix(w.EndTime, 0)
```

Comment:
Since we don't store timezone data you might want to explicity return the time in UTC and not the server timezone. It's the same but maybe less confuising for people? `time.Unix(w.StartTime, 0).UTC()`

Reply:
Good call